src/navigation/waypoint.py

```diff
@@ -73,36 +41,36 @@ def evaluate(self, ud) -> str:
         :param ud:  State machine user data
         :return:    Next state
         """
-        current_waypoint = self.current_waypoint(ud)
+        current_waypoint = self.context.course.current_waypoint()
         if current_waypoint is None:
             return "done"
 
         # Go into the single fiducial state if we see it early
-        if current_waypoint.fiducial_id != NO_FIDUCIAL and self.current_fid_pos(ud) is not None:
+        if current_waypoint.fiducial_id != Environment.NO_FIDUCIAL and self.context.env.current_fid_pos() is not None:
             return "single_fiducial"
 
+        # Attempt to find the waypoint in the TF tree and drive to it
         try:
-            waypoint_pos = self.waypoint_pose(ud.waypoint_index).position_vector()
+            waypoint_pos = self.context.course.current_waypoint_pose().position_vector()
             cmd_vel, arrived = get_drive_command(
                 waypoint_pos,
-                self.context.get_rover_pose(),
+                self.context.rover.get_pose(),
                 STOP_THRESH,
                 DRIVE_FWD_THRESH,
             )
             if arrived:
                 if current_waypoint.fiducial_id == NO_FIDUCIAL:
                     # We finished a regular waypoint, go onto the next one
-                    ud.waypoint_index += 1
+                    self.context.course.inc_waypoint()
                 else:
                     # We finished a waypoint associated with a fiducial id, but we have not seen it yet.
-                    return "single_fiducial"
-            self.context.drive_command(cmd_vel)
+                    return "search"
+            self.context.rover.send_drive_command(cmd_vel)
         except (
             tf2_ros.LookupException,
             tf2_ros.ConnectivityException,
             tf2_ros.ExtrapolationException,
         ):
-            # TODO: probably go into some waiting state
```

Comment:
I'm confused, should we keep this in for later or not?

Reply:
imo, no. TF tree lookup failure is fairly common, the buffer update isn't instantaneous so it takes a few milliseconds after publishing to lookup without exceptions. In other words, getting an exception should be considered normal operation. If whoever wrote that comment disagrees we can figure it out though (@qhdwight ?)

Reply:
Yes missing transform updates should be considered "normal," as in we should handle them properly.